### PR TITLE
Ee 4469/implment cat b non salaried

### PR DESCRIPTION
**Pull request by mikehalmaho**

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/proving/income/validator/AnnualisedAverageCalculator.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+    private static final Integer MONTHS_PER_YEAR = 12;
+
+    public static BigDecimal calculate(Map<Integer, BigDecimal> aggregatedMonthlyIncome) {
+        if(aggregatedMonthlyIncome.size() == 0) {
+            return new BigDecimal("0.00");
+        }
+
+        BigDecimal total = aggregatedMonthlyIncome.values().stream().reduce(BigDecimal.ZERO, BigDecimal::add);
+
```
Comment (mikehalmaho) [inline src/main/java/uk/gov/digital/ho/proving/income/validator/AnnualisedAverageCalculator.java:17]: Shouldn't we ignore any months with zero earnings?  What about negative earnings?

**Review by CameronDowner — APPROVED** [review]
Over the shoulder review with Mike - we raised a few minor points around smells but logic is sound and test coverage is good.
Approved for merge

File: src/main/java/uk/gov/digital/ho/proving/income/api/FinancialStatusService.java [hunk 2]
```diff
@@ -82,20 +83,41 @@ public FinancialStatusCheckResponse getFinancialStatus(@Valid @RequestBody Finan
...
+        IncomeRecord applicantIncomeRecord = hmrcClient.getIncomeRecord(
             new Identity(mainApplicant.forename(), mainApplicant.surname(), mainApplicant.dateOfBirth(), sanitisedNino),
             startSearchDate,
             request.applicationRaisedDate());
 
-        Individual individual = individualFromRequestAndRecord(request, incomeRecord.hmrcIndividual(), sanitisedNino);
+        incomeRecords.put(individualFromRequestAndRecord(request, applicantIncomeRecord.hmrcIndividual(), sanitisedNino), applicantIncomeRecord);
 
-        FinancialStatusCheckResponse response = calculateResponse(request.applicationRaisedDate(), request.dependants(), startSearchDate, incomeRecord, individual);
+        if (request.applicants().size() > 1) {
```
Comment (mikehalmaho) [inline src/main/java/uk/gov/digital/ho/proving/income/api/FinancialStatusService.java:107]: What if there are more than 2 applicants?  Return an error?

File: src/main/java/uk/gov/digital/ho/proving/income/validator/AnnualisedAverageCalculator.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package uk.gov.digital.ho.proving.income.validator;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.Map;
+
+public class AnnualisedAverageCalculator {
```
Comment (mikehalmaho) [inline src/main/java/uk/gov/digital/ho/proving/income/validator/AnnualisedAverageCalculator.java:7]: What does Annualised Average even mean?  Better to call this ProjectedAnnualIncomeCalculator?

File: src/test/java/uk/gov/digital/ho/proving/income/validator/CatBNonSalariedIncomeValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+import java.time.LocalDate;
+import java.time.Month;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.gov.digital.ho.proving.income.validator.CatASalariedTestData.getDate;
+import static uk.gov.digital.ho.proving.income.validator.CatBNonSalariedTestData.*;
+
+public class CatBNonSalariedIncomeValidatorTest {
```
Comment (mikehalmaho) [inline src/test/java/uk/gov/digital/ho/proving/income/validator/CatBNonSalariedIncomeValidatorTest.java:19]: This has good coverage around various different scenarios, we could do with one big nasty test that covers off several scenarios to make sure that they work in tandem.  Just the one though...

File: src/test/java/steps/StepAssertor.java [hunk 5]
```diff
@@ -76,28 +77,41 @@ private static boolean testCategoryCheck(String json, String responseObject, Str
...
+    private static boolean testCategoryCheckNinos(String json, String responseObject, String key, String expected) {
+        if (responseObject.startsWith("Category ") && key.contains("nino")) {
+            String jsonPath = "$.categoryChecks..[?(@.calculationType == '" + responseObject + "')].individuals..[?(@.nino == '" + expected + "')]";
+            JSONArray jsonData = read(json, jsonPath);
+            assertTrue(jsonData.size() > 0, String.format("%s %s: Unable to find category check nino %s\nJSON: %s", responseObject, key, expected, json));
+            return true;
+        }
+        return false;
+    }
+
```
Comment (mikehalmaho) [inline src/test/java/steps/StepAssertor.java:109]: This is hard to follow.  In general this class is becoming a bit bloated and difficult to understand - e.g. what do those complicated json paths even mean?  This could do with refactoring at some point soon-ish.
